eppo_core/src/event_ingestion/event_ingestion.rs

```diff
@@ -0,0 +1,228 @@
+        ));
+        runtime.spawn_untracked(delivery::delivery(
+            delivery_uplink,
+            delivery_downlink,
+            delivery_status_tx.clone(),
+            config.max_retries,
+            event_delivery,
+        ));
+        runtime.spawn_untracked(retry::retry(
+            retry_uplink,
+            // This is sending to input of delivery
+            batcher_downlink,
+            config.max_retries,
+            config.retry_interval,
+            config.max_retry_delay,
+        ));
+
+        // For now, nobody is interested in delivery statuses.
+        let _ = delivery_status_rx;
+
+        EventIngestion { tx: input }
+    }
+
+    pub fn track(&self, event_type: String, payload: serde_json::Value) {
+        let event = Event {
+            uuid: Uuid::new_v4(),
+            timestamp: crate::timestamp::now(),
+            event_type,
+            payload,
+        };
+
+        self.track_event(event);
+    }
+
+    fn track_event(&self, event: Event) {
+        let result = self
+            .tx
+            .try_send(BatchedMessage::singleton(QueuedEvent::new(event)));
+
+        if let Err(err) = result {
+            log::warn!(target: "eppo", "failed to submit event to event ingestion: {}", err);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::event_ingestion::event::Event;
+    use crate::timestamp::now;
+    use serde_json::json;
+    use tokio::time::Duration;
+    use uuid::Uuid;
+    use wiremock::matchers::{body_json, header, method, path};
+    use wiremock::{Mock, MockServer, ResponseTemplate};
+
+    fn init() {
+        let _ = env_logger::try_init();
+    }
+
+    #[tokio::test]
+    async fn test_dispatch_starts_delivery() {
+        init();
+
+        let event = new_test_event();
+
+        let mock_server = MockServer::start().await;
+        Mock::given(method("POST"))
+            .and(path("/"))
+            .and(body_json(&json!({
+                "eppo_events": [event.clone()],
+            })))
+            .and(header("x-eppo-token", "test-sdk-key"))
+            .respond_with(ResponseTemplate::new(200).set_body_json(json!({
+                "failed_events": []
+            })))
+            .expect(1)
+            .mount(&mock_server)
+            .await;
+
+        run_dispatcher_task(event.clone(), &mock_server.uri()).await;
+
+        mock_server.verify().await;
+    }
+
+    #[tokio::test]
+    async fn test_dispatch_failed_after_max_retries() {
+        init();
+
+        let event = new_test_event();
+
+        let mock_server = MockServer::start().await;
+        Mock::given(method("POST"))
+            .and(path("/"))
+            .and(body_json(&json!({"eppo_events": [event] })))
+            .and(header("x-eppo-token", "test-sdk-key"))
+            .respond_with(ResponseTemplate::new(200).set_body_json(json!({
+                "failed_events": [event.uuid],
+            })))
+            .expect(3) // 1 regular attempt + 2 retries
```

Comment:
nice but i kinda miss introspecting into the internal state of the channels. Consider exposing the received like before to keep that, unless you have a strong reason not to

Reply:
The main issue is that nobody is reading that channel in production code yet, so it clogs and blocks the sending part, so everything halts.

Also, because it's a single-consumer channel, we won't be able to read it in tests as soon as we introduce persistence.

I agree that it's useful for testing but the tests that rely on it _are_ peeking into implementation details, so shall be lower-level tests (e.g., arrange a couple of components and test that they work as expected).